### PR TITLE
Z-star vertical coordinate

### PR DESCRIPTION
**Pull request by simone-silvestri**
This PR introduces a ZStar, free surface following vertical coordinate. 

the vertical coordinate evolves in time and is implemented following the recipe in 
[Rescaled height coordinates for accurate representation of free-surface flows in ocean circulation models ](https://www.sciencedirect.com/science/article/abs/pii/S1463500303000544) that allows the implementation of a general vertical coordinate where the only specification needed is the vertical derivative of the vertical coordinate.

Closes https://github.com/CliMA/Oceananigans.jl/issues/1679

### CODE REVIEW

**Comment by simone-silvestri** [comment]
global tracer conservation for the hydrostatic lock-release configuration in `validation/moving_coordinate/z_star_lock_release.jl`. Better than the z-coordinate but still not machine-precision conserving.

Screenshot 2023-12-21 at 11 00 02 AM

**Comment by glwagner** [comment]
Please describe your design in words, the challenges that exist, and how you plan to meet those challenges.

**Comment by glwagner** [comment]
There are two things at play here. First, there is a refactoring of the equation set to allow for a generalized vertical coordinate. This actually involves rewriting the equations. To merge that generalization, I think we would also like to see the new equations that are being solved written up in the docs. When the equations are generalized, this would also involve implementing a new abstraction for the vertical coordinate, along with a `vertical_coordinate = ZVerticalCoordinate()` default, which preserves existing behavior.

The second PR should then implement a `ZStarVerticalCoordinate()` (other better names are fine too).

So to me this feels like we need two PRs, one with docs update and API design, and then a second with z* implementation.

File: src/Models/HydrostaticFreeSurfaceModels/z_star_vertical_spacing.jl [hunk 1]
```diff
@@ -33,7 +33,7 @@ function GeneralizedSpacingGrid(grid::AbstractUnderlyingGrid{FT, TX, TY, TZ}, ::
     
     # Initial "at-rest" conditions
     launch!(architecture(grid), grid, :xy, _update_zstar!, 
-            sⁿ, s⁻, ΔzF, ΔzC, ZeroField(grid), grid, Val(grid.Nz))
+            sⁿ, s⁻, ΔzF, ΔzC, ZeroField(), grid, Val(grid.Nz))
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/z_star_vertical_spacing.jl:36]: Should we use Oceananigans notation for `Δz`?

File: src/Models/HydrostaticFreeSurfaceModels/HydrostaticFreeSurfaceModels.jl [hunk 2]
```diff
@@ -86,6 +86,10 @@ Return a flattened `NamedTuple` of the prognostic fields associated with `Hydros
 @inline prognostic_fields(model::HydrostaticFreeSurfaceModel) =
     hydrostatic_prognostic_fields(model.velocities, model.free_surface, model.tracers)
 
+@inline prognostic_fields(model::HydrostaticFreeSurfaceModel{<:Any, <:Any, <:Any, <:Any, ZStarSpacingGrid}) =
+    merge(hydrostatic_prognostic_fields(model.velocities, model.free_surface, model.tracers), 
+          (; ∂t_∂s = grid.Δzᵃᵃᶠ.∂t_∂s))
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/HydrostaticFreeSurfaceModels.jl:91]: What does `∂t_∂s` mean?

Comment (glwagner) [reply src/Models/HydrostaticFreeSurfaceModels/HydrostaticFreeSurfaceModels.jl:91]: You probably want to define an interface to `grid.Δzᵃᵃᶠ.∂t_∂s`. This is specific to the grids we have because it requires the property `Δzᵃᵃᶠ` to exist. We discourage writing assumptive code like this.

Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/HydrostaticFreeSurfaceModels.jl:91]: right, I will add that (it actually already exists, I just forgot to add it there)

File: src/ImmersedBoundaries/ImmersedBoundaries.jl [hunk 3]
```diff
@@ -150,8 +150,11 @@ Adapt.adapt_structure(to, ibg::IBG{FT, TX, TY, TZ}) where {FT, TX, TY, TZ} =
                                      nothing, 
                                      nothing)
 
-with_halo(halo, ibg::ImmersedBoundaryGrid) =
-    ImmersedBoundaryGrid(with_halo(halo, ibg.underlying_grid), ibg.immersed_boundary)
+function with_halo(halo, ibg::ImmersedBoundaryGrid)
+    new_grid = with_halo(halo, ibg.underlying_grid)
+    active_cells_map = !isnothing(ibg.interior_active_cells)
+    return ImmersedBoundaryGrid(new_grid, ibg.immersed_boundary; active_cells_map)
+end
```
Comment (glwagner) [inline src/ImmersedBoundaries/ImmersedBoundaries.jl:157]: does this have anything to do with the present PR?

File: src/ImmersedBoundaries/ImmersedBoundaries.jl [hunk 4]
```diff
@@ -238,6 +241,10 @@ As well as
 @inline immersed_inactive_node(i, j, k, ibg::IBG, LX, LY, LZ) =  inactive_node(i, j, k, ibg, LX, LY, LZ) &
                                                                 !inactive_node(i, j, k, ibg.underlying_grid, LX, LY, LZ)
 
+# Underlying grids are never immersed!                                                                
+@inline immersed_peripheral_node(i, j, k, grid::AbstractUnderlyingGrid, LX, LY, LZ) = false
+@inline   immersed_inactive_node(i, j, k, grid::AbstractUnderlyingGrid, LX, LY, LZ) = false
```
Comment (glwagner) [inline src/ImmersedBoundaries/ImmersedBoundaries.jl:246]: relation to present PR?

File: src/Grids/rectilinear_grid.jl [hunk 5]
```diff
@@ -19,29 +19,31 @@ struct RectilinearGrid{FT, TX, TY, TZ, FX, FY, FZ, VX, VY, VZ, Arch} <: Abstract
...
+                                                                          TX, TY, TZ,
+                                                                          FX, VX, FY,
+                                                                          VY, FZF, FZC, VZ}
+                                                                                           
+        return new{FT, TX, TY, TZ, FX, FY, FZF, FZC, VX, VY, VZ, Arch}(arch, Nx, Ny, Nz,
+                                                                       Hx, Hy, Hz, Lx, Ly, Lz, 
+                                                                       Δxᶠᵃᵃ, Δxᶜᵃᵃ, xᶠᵃᵃ, xᶜᵃᵃ,
+                                                                       Δyᵃᶠᵃ, Δyᵃᶜᵃ, yᵃᶠᵃ, yᵃᶜᵃ,
+                                                                       Δzᵃᵃᶠ, Δzᵃᵃᶜ, zᵃᵃᶠ, zᵃᵃᶜ)
+    end
```
Comment (glwagner) [inline src/Grids/rectilinear_grid.jl:46]: This could be its own PR.

File: src/ImmersedBoundaries/grid_fitted_bottom.jl [hunk 6]
```diff
@@ -8,6 +8,7 @@ using Oceananigans.BoundaryConditions: FBC
 using Printf
 
 import Oceananigans.TurbulenceClosures: z_bottom
+import Oceananigans.Architectures: on_architecture
```
Comment (glwagner) [inline src/ImmersedBoundaries/grid_fitted_bottom.jl:11]: relation to present PR?

File: src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+spacings for a generalized vertical coordinate system. The reference (non-moving) spacings are stored in `Δr`. 
+`Δ` contains the spacings associated with the moving coordinate system.
+`s⁻`, `sⁿ` and `∂t_∂s` fields are the vertical derivative of the vertical coordinate (∂Δ/∂Δr)
+at timestep `n-1` and `n` and it's time derivative.
+`denomination` contains the "type" of generalized vertical coordinate (the only one implemented is `ZStar`)
+"""
+abstract type AbstractVerticalSpacing{R} end
+
+const AbstractVerticalSpacingRG  = RectilinearGrid{<:Any, <:Any, <:Any, <:Any, <:Any, <:Any, <:AbstractVerticalSpacing}
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl:27]: why is the property all the way at the end of the list of properties of the grid? makes this harder to read

File: src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+retrieve_static_grid(grid) = grid
+
+function retrieve_static_grid(grid::AbstractVerticalSpacingImmersedGrid) 
+    underlying_grid  = retrieve_static_grid(grid.underlying_grid)
+    active_cells_map = !isnothing(grid.interior_active_cells)
+    return ImmersedBoundaryGrid(underlying_grid, grid.immersed_boundary; active_cells_map)
+end
+
+reference_Δzᵃᵃᶠ(grid) = grid.Δzᵃᵃᶠ
+reference_Δzᵃᵃᶜ(grid) = grid.Δzᵃᵃᶜ
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl:48]: does zstar not work for a deep sphere case?

File: src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+
+generalized_spacing_grid(grid, coord) = grid
+update_vertical_spacing!(model, grid; kwargs...) = nothing
+
+##### 
+##### Vertical spacings for a generalized vertical coordinate system
+#####
+
+# Very bad for GPU performance!!! (z-values are not coalesced in memory for z-derivatives anymore)
+# TODO: make z-direction local in memory by not using Fields
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl:107]: This comment makes me worried. There are better ways to make the z-direction local; the `Field` abstraction has nothing to do with it. Remember that `getindex` is divorced from the memory layout. The 3rd index can be local. The thing that has to change is simply how the arrays are constructed via `newdata` in the grids module.

File: src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_model.jl [hunk 10]
```diff
@@ -100,30 +101,43 @@ Keyword arguments
...
+                                   tracers = (:T, :S),
+                       forcing::NamedTuple = NamedTuple(),
+                                   closure = nothing,
+           boundary_conditions::NamedTuple = NamedTuple(),
+             particles::ParticlesOrNothing = nothing,
+     biogeochemistry::AbstractBGCOrNothing = nothing,
+                                velocities = nothing,
+                                  pressure = nothing,
+                        diffusivity_fields = nothing,
+           generalized_vertical_coordinate = nothing,
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_model.jl:123]: `generalized` is implied fine here, so I would go with the name `vertical_coordinate`.

File: src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl [hunk 11]
```diff
@@ -82,10 +83,11 @@ implicitly during time-stepping.
     
     model_fields = merge(hydrostatic_fields(velocities, free_surface, tracers), auxiliary_fields)
 
-    return ( - U_dot_∇v(i, j, k, grid, advection, velocities)
+    return ( - U_dot_∇v(i, j, k, grid, advection, velocities) 
              - explicit_barotropic_pressure_y_gradient(i, j, k, grid, free_surface)
              - y_f_cross_U(i, j, k, grid, coriolis, velocities)
              - ∂yᶜᶠᶜ(i, j, k, grid, hydrostatic_pressure_anomaly)
+             - grid_slope_contribution_y(i, j, k, grid, free_surface, buoyancy, model_fields)
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl:90]: is this a pressure contribution? I can't quite tell

File: src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl (hunk 8 above)
Comment (glwagner) [reply src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl:48]: it's unclear to me whether this is a "reference" or "resting"

File: src/Models/HydrostaticFreeSurfaceModels/z_star_vertical_spacing.jl [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+    sᶜᶜⁿ :: SCC
+    sᶠᶜⁿ :: SFC
+    sᶜᶠⁿ :: SCF
+    sᶠᶠⁿ :: SFF
+    sᶜᶜ⁻ :: SCC
+ ∂t_∂s   :: SCC
+end
+
+Adapt.adapt_structure(to, coord::ZStarSpacing) = 
+            ZStarSpacing(Adapt.adapt(to, coord.Δr),
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/z_star_vertical_spacing.jl:36]: is that an 8 character indent

File: src/Models/HydrostaticFreeSurfaceModels/z_star_vertical_spacing.jl [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+The vertical spacing is defined by a reference spacing `Δr` and a scaling `s` that obeys
+```math
+s = (η + H) / H
+```
+where ``η`` is the free surface height and ``H`` the vertical depth of the water column
+
+# Fields
+- `Δr`: reference vertical spacing with `η = 0`
+- `sⁿ`: scaling of the vertical coordinate at time step `n`
+- `s⁻`:scaling of the vertical coordinate at time step `n - 1`
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/z_star_vertical_spacing.jl:22]:
```suggestion
- `s⁻`: scaling of the vertical coordinate at time step `n - 1`
```

File: src/Models/HydrostaticFreeSurfaceModels/z_star_vertical_spacing.jl [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+A vertical spacing for the hydrostatic free surface model that follows the free surface.
+The vertical spacing is defined by a reference spacing `Δr` and a scaling `s` that obeys
+```math
+s = (η + H) / H
+```
+where ``η`` is the free surface height and ``H`` the vertical depth of the water column
+
+# Fields
+- `Δr`: reference vertical spacing with `η = 0`
+- `sⁿ`: scaling of the vertical coordinate at time step `n`
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/z_star_vertical_spacing.jl:21]: Are there any options other than `s`? It doesn't produce the most obvious expressions and also `s` is used for salinity.

File: src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+@inline Δzᶜᶠᶠ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑyᵃᶠᵃ(i, j, k, grid, Δzᶜᶜᶠ)
+@inline Δzᶜᶠᶜ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑyᵃᶠᵃ(i, j, k, grid, Δzᶜᶜᶜ)
+
+@inline Δzᶠᶜᶠ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑxᶠᵃᵃ(i, j, k, grid, Δzᶜᶜᶠ)
+@inline Δzᶠᶜᶜ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑxᶠᵃᵃ(i, j, k, grid, Δzᶜᶜᶜ)
+
+@inline Δzᶠᶠᶠ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑxyᶠᶠᵃ(i, j, k, grid, Δzᶜᶜᶠ)
+@inline Δzᶠᶠᶜ(i, j, k, grid::AbstractVerticalSpacingGrid) = ℑxyᶠᶠᵃ(i, j, k, grid, Δzᶜᶜᶜ)
+
+@inline Δz_reference(i, j, k, Δz::Number) = Δz
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl:119]: I'm not a huge fan of this notation because it mixes symbols (`Δz`) with language (`reference`)

File: src/Grids/rectilinear_grid.jl [hunk 16]
```diff
@@ -1,4 +1,4 @@
-struct RectilinearGrid{FT, TX, TY, TZ, FX, FY, FZ, VX, VY, VZ, Arch} <: AbstractUnderlyingGrid{FT, TX, TY, TZ, Arch}
+struct RectilinearGrid{FT, TX, TY, TZ, FX, FY, FZF, FZC, VX, VY, VZ, Arch} <: AbstractUnderlyingGrid{FT, TX, TY, TZ, Arch}
```
Comment (glwagner) [inline src/Grids/rectilinear_grid.jl:1]: Why do FX, FY take priority over FZF and FZC (not to mention Arch) here?

File: src/Advection/vector_invariant_cross_upwinding.jl [hunk 17]
```diff
@@ -21,16 +21,19 @@
     @inbounds û = u[i, j, k]
     δ_stencil = scheme.upwinding.divergence_stencil
 
-    δᴿ = _biased_interpolate_xᶠᵃᵃ(i, j, k, grid, scheme, scheme.divergence_scheme, bias(û), flux_div_xyᶜᶜᶜ, δ_stencil, u, v) 
+    δᴿ  =    _biased_interpolate_xᶠᵃᵃ(i, j, k, grid, scheme, scheme.divergence_scheme, bias(û), flux_div_xyᶜᶜᶜ, δ_stencil, u, v) 
+    ∂ts = _symmetric_interpolate_xᶠᵃᵃ(i, j, k, grid, scheme, cross_scheme, V_times_∂t_∂s_grid)
```
Comment (glwagner) [inline src/Advection/vector_invariant_cross_upwinding.jl:25]: Is `∂t_∂s` right? Or is this just `∂t_s`? Note, this should be consistent everywhere. (I see a `∂ts` below.) We use the form `∂t_s` as a shorthand for `∂s/∂t` (the time-derivative of `s` with respect to `t`), elsewhere we have written `∂z_u` for example.

File: src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl (hunk 9 above)
Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl:107]: this is actually superseded by the fact that we do not allocate anymore a 3d field but compute \Delta z locally by doing \Delta r * s where s is a two-dimensional field, so memory access should be coalesced now

File: src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl (hunk 11 above)
Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_tendency_kernel_functions.jl:90]: yep, it's the hydrostatic pressure contribution coming from the fact that the grid is sloped. What about `grid_slope_hydrostatic_pressure_y`?

File: src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl [hunk 18]
```diff
@@ -116,23 +116,27 @@ update_vertical_spacing!(model, grid; kwargs...) = nothing
...
 
-@inline Δz_reference(i, j, k, Δz::AbstractVerticalSpacing{<:Number}) = Δz.Δr
-@inline Δz_reference(i, j, k, Δz::AbstractVerticalSpacing) = @inbounds Δz.Δr[k]
+@inline Δr(i, j, k, Δz::AbstractVerticalSpacing{<:Number}) = Δz.Δr
+@inline Δr(i, j, k, Δz::AbstractVerticalSpacing) = @inbounds Δz.Δr[k]
 
-@inline Δzᶜᶜᶠ_reference(i, j, k, grid) = Δz_reference(i, j, k, grid.Δzᵃᵃᶠ)
-@inline Δzᶜᶜᶜ_reference(i, j, k, grid) = Δz_reference(i, j, k, grid.Δzᵃᵃᶜ)
+@inline Δrᶜᶜᶠ(i, j, k, grid) = Δr(i, j, k, grid.Δzᵃᵃᶠ)
+@inline Δrᶜᶜᶜ(i, j, k, grid) = Δr(i, j, k, grid.Δzᵃᵃᶜ)
```
Comment (glwagner) [inline src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl:130]: this is better!

I would like to know, does this mean we use the opposite notation as MOM6? I believe in MOM6 `r` is the generalized vertical coordinate and `z` is the height coordinate. It would seem with these definitions that `z` is the name of the "generalized" coordinate and `r` is the reference (which at least for z* is the height coordinate, though for other coordinates I suppose it might be something else).

Comment (simone-silvestri) [reply src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl:130]: I was thinking the reference `r` is the `z*` (always fixed between 0 and -H), while the z is the moving grid defined between -H and eta. We probably need to also add an `rnode` for this grid.

Comment (glwagner) [reply src/Models/HydrostaticFreeSurfaceModels/generalized_vertical_spacing.jl:130]: Yes for sure. It's a notation battle. `rnode` implies that "r" means the "third coordinate" (not specifically the reference coordinate).

It'd require more herculean effort for us to be able to denote "z" as "height", rather than the generalized vertical coordinate, though.

Still something to ponder. It's not impossible. I don't think we need it here.

"r" for reference makes a lot of sense. Perhaps, we can choose a different name for `rnode`. Like `ζnode` or something.

File: src/Advection/vector_invariant_cross_upwinding.jl (hunk 17 above)
Comment (simone-silvestri) [reply src/Advection/vector_invariant_cross_upwinding.jl:25]: ok make sense, I will switch everything to `∂t_s`. Then we can think about a better name for `s`. In MITgcm they use `hfac`. In NEMO they use `e3`.

File: src/Grids/rectilinear_grid.jl (hunk 16 above)
Comment (simone-silvestri) [reply src/Grids/rectilinear_grid.jl:1]: I don't think we ever dispatch on Arch. FZF and FZC can definitely take priority. I ll open a new PR to do that.

**Comment by glwagner** [comment]
@simone-silvestri discussed a bit the user interface for building models with a z* vertical coordinate. Here's a prototype design we came up with:

```julia
reference_levels = exponential_z_faces(Nz=40, depth=6000)

Nx = 1440
Ny = 600
Nz = length(reference_levels) - 1

grid = LatitudeLongitudeGrid(arch;
                             size = (Nx, Ny, Nz),
                             halo = (7, 7, 7),
                             z = ZStarVerticalCoordinate(reference_levels),
                             latitude  = (-75, 75),
                             longitude = (0, 360))
```

We like this because with z*, the znodes, and zspacings are always changing. We think this makes that semantically clear.

Note there are some challenges for output that we have to solve, because the grid has to be updated to do computations. But I think we can solve these with `FieldTimeSeries`.

**Comment by glwagner** [comment]
We can break some of this up into at least 3 PRs:

1. Compute "numerical bottom height" when building immersed boundary grid
2. Implement a nonlinear free surface
3. Do z*

**Comment by simone-silvestri** [comment]
superseded by #3956
